// wtd-core/dist/offscreen/helper.js
import { WorkerTaskMessage } from '../WorkerTaskMessage.js';
import { OffscreenPayload } from './OffscreenPayload.js';
import { OffscreenWorkerCommandRequest, OffscreenWorkerCommandResponse } from './OffscreenWorker.js';
export const getOffscreenCanvas = (offScreenPayload) => {
    return offScreenPayload ? offScreenPayload.message.drawingSurface : undefined;
};
export const initOffscreenCanvas = async (workerTask, canvas) => {
    const offscreenCanvas = canvas.transferControlToOffscreen();
    const offscreenPayloadRenderer = new OffscreenPayload({
        drawingSurface: offscreenCanvas,
        width: canvas.clientWidth,
        height: canvas.clientHeight,
        pixelRatio: window.devicePixelRatio
    });
    await workerTask.sentMessage({
        message: WorkerTaskMessage.fromPayload(offscreenPayloadRenderer, OffscreenWorkerCommandRequest.INIT_OFFSCREEN_CANVAS),
        transferables: [offscreenCanvas],
        awaitAnswer: true,
        answer: OffscreenWorkerCommandResponse.INIT_OFFSCREEN_CANVAS_COMPLETE
    });
};
export const recalcAspectRatio = (canvas, clientWidth, clientHeight) => {
    canvas.width = canvas.height * (clientWidth / clientHeight);
    return canvas.width;
};
