// wtd-core/dist/utilities.js
import { RawPayload } from './RawPayload.js';
import { WorkerTaskMessage } from './WorkerTaskMessage.js';
import { WorkerTaskCommandRequest, WorkerTaskCommandResponse } from './WorkerTaskWorker.js';
export const fillTransferables = (buffers, transferables, cloneBuffers) => {
    for (const buffer of buffers) {
        const potentialClone = cloneBuffers ? buffer.slice(0) : buffer;
        const outputBuffer = potentialClone.buffer;
        if (outputBuffer) {
            transferables.push(outputBuffer);
        }
        else {
            transferables.push(potentialClone);
        }
    }
};
/**
 * Applies values from parameter object via set functions or via direct assignment.
 *
 * @param {object} objToAlter The objToAlter instance
 * @param {AssociatedArrayType} params The parameter object
 * @param {boolean} forceCreation Force the creation of a property
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const applyProperties = (objToAlter, params, forceCreation) => {
    if (!params)
        return;
    for (const [k, v] of Object.entries(params)) {
        const funcName = 'set' + k.substring(0, 1).toLocaleUpperCase() + k.substring(1);
        if (Object.prototype.hasOwnProperty.call(objToAlter, funcName) && typeof objToAlter[funcName] === 'function') {
            objToAlter[funcName] = v;
        }
        else if (Object.prototype.hasOwnProperty.call(objToAlter, k) || forceCreation) {
            objToAlter[k] = v;
        }
    }
};
export const createWorkerBlob = (code) => {
    const simpleWorkerBlob = new Blob(code, { type: 'application/javascript' });
    return window.URL.createObjectURL(simpleWorkerBlob);
};
export const initChannel = async (workerOne, workerTwo) => {
    const channel = new MessageChannel();
    const promises = [];
    const payloadOne = new RawPayload({
        port: channel.port1
    });
    promises.push(workerOne.sentMessage({
        message: WorkerTaskMessage.fromPayload(payloadOne, WorkerTaskCommandRequest.INIT_CHANNEL),
        transferables: [channel.port1],
        awaitAnswer: true,
        answer: WorkerTaskCommandResponse.INIT_CHANNEL_COMPLETE
    }));
    const payloadTwo = new RawPayload({
        port: channel.port2
    });
    promises.push(workerTwo.sentMessage({
        message: WorkerTaskMessage.fromPayload(payloadTwo, WorkerTaskCommandRequest.INIT_CHANNEL),
        transferables: [channel.port2],
        awaitAnswer: true,
        answer: WorkerTaskCommandResponse.INIT_CHANNEL_COMPLETE
    }));
    return Promise.all(promises);
};
